### PR TITLE
Added spec monitor_ceph_pool_utilization.adoc

### PR DESCRIPTION
**Pull request by shtripat**
Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc]: Proposed changes can be:

* Introduce a collectd plugin which collects the ceph pool utilization data from underlying cluster.
* Introduce a template to configure the above mentioned collectd plugin with details such as cluster name and thresholds.
* Introduce threshold handlers on the alerting module to custom handle the pool(and cluster as well as ceph df itself provides both) utilization threshold breach.

----
Note:
* Any stat that is collected by collectd is guaranteed to land in time series db by the collectd plugin configured as write plugin(in our case the one configured to write to time series db).
* The collectd plugin to capture pool utilization will be configured on the leader mon node in the cluster at the time of creation of cluster. The removal of pool does not require the removal of configuration of pool utilization plugin as it is auto handled by ceph in case of using ceph df command. However the removal of cluster MIGHT require removal of respective collectd configuration.
----

Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc]: Ok. will change accordingly

Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc]: Also I think it would be better to mention the command or approach taken in the plugin to collect usage stats because, as you have mentioned below there are many techniques available to achieve this task we probably need to evaluate the flaws of each of them and make a final call. 1 problem that I have seen with ceph df with json format option is it provides only used size and % used but not total. I remember hearing that it'll be easier to add it on ceph side but we might need to raise an issue and track it if we go with that approach.

**Comment by shtripat** [comment]
@anmolbabu can you have a look at the https://github.com/rochaporto/collectd-ceph/tree/master/plugins
This plugin as well uses json output and pushes the details to time series DB.
I understand that in case of json we would have to calculate the percentage usage value but with suggestion from ceph team on https://bugzilla.redhat.com/show_bug.cgi?id=1360230, isnt it doable.

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc:99]: Probably, this needs to be either of the below:

1. The pool object in tendrl ceph integration definitions, will provide a way to indicate a way to read the utilization data from some directory in etcd if monitoring is enabled and the tendrl-api will look into this path to provide the utilization data
or
2. The performance-monitoring application will periodically update the utilization data to a location in etcd and also the performance monitoring application will expose apis to access this data and the tendrl-api will proxy to it.

I personally think that approach 2 although is a 2-step process, is safer and better than approach 1 of involving changes to definitions on the fly once the monitoring is enabled.

**Comment by shtripat** [comment]
>Probably, this needs to be either of the below:
>
>    The pool object in tendrl ceph integration definitions, will provide a way to indicate a way to read the >utilization data from some directory in etcd if monitoring is enabled and the tendrl-api will look into this path >to provide the utilization data
>    or
>    The performance-monitoring application will periodically update the utilization data to a location in etcd >and also the performance monitoring application will expose apis to access this data and the tendrl-api will >proxy to it.

>I personally think that approach 2 although is a 2-step process, is safer and better than approach 1 of >involving changes to definitions on the fly once the monitoring is enabled.

Will change the spec as suggested

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc:99]: As we discussed this goes as part of the pool entity itself. And since the update of utilization is anyway being done as part of pool entity and is regularly synced to etcd, by the ceph-integration, we no more require this api from performance-monitoring

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc]: Typo here: 'The'

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc]: Typo here : 'ceph' ?

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc]: "And the plugin will be configured once a cluster is created And removal of pools has no impact on the plugin or its configuration. However, removal of the cluster will handle the removal of configuration of this plugin. The details regarding cluster removal will however be covered in a separate spec as it is beyond the scope of this spec."

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc:146]: 1 problem that I have seen with ceph df with json format option is it doesn't provide %used details and total capacity of pool. I remember hearing that it'll be easier to add it on ceph side but we might need to raise an issue and track it if we go with that approach. Also in normal cli command without json o/p format, it provides used percentage but not total but we'll need to parse the command output as a string.

File: specs/monitor_ceph_pool_utilization.adoc
Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc:99]: ok got it. will remove this

File: specs/monitor_ceph_pool_utilization.adoc
Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc]: Done

File: specs/monitor_ceph_pool_utilization.adoc
Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc]: Done

File: specs/monitor_ceph_pool_utilization.adoc
Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc:146]: I rember in skyring it was done using the non-json output. Will try to the same thing.

File: specs/monitor_ceph_pool_utilization.adoc
Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc]: ok

**Review by brainfunked — COMMENTED** [review]
Looks good, will approve once the changes requested by Anmol have been made.

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc:159]: The plugin needs to be configured only on the leader mon node(only 1 node) because if it it is running on multiple nodes of a cluster, all of them will raise threshold breach alerts(if any) for the same set of pools in cluster and hence there will be as many duplicate alerts as the number of mons in cluster(if deployed on all mons) which can be handled internally(if un-avoidable) but it might be inefficient.

File: specs/monitor_ceph_pool_utilization.adoc
Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc:146]: We should be using non json output of `ceph df` command and use here. I feel @anmolbabu already has a patch for this.

File: specs/monitor_ceph_pool_utilization.adoc
Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc:159]: We are thinking of having some kind of tagging mechanism for the nodes and using that the leader mon can have this enabled.
@brainfunked what do you feel? Should we keep this requirement here at the moment or later we would take care.

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [reply specs/monitor_ceph_pool_utilization.adoc:146]: @shtripat I have a patch for collectd plugin @https://github.com/Tendrl/ceph_integration/pull/11

File: specs/monitor_ceph_pool_utilization.adoc
Comment (anmolbabu) [inline specs/monitor_ceph_pool_utilization.adoc:172]: I think you probably need to create issues on associated components and link it here not the spec issue(I feel)

**Review by anmolbabu — COMMENTED** [review]
1 minor comment and 1 answer awaited from @brainfunked

File: specs/monitor_ceph_pool_utilization.adoc
Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc:172]: Different module specific issues actually I keep under references and spec issue here :) thats what was followed in earlier specs.

Comment (anmolbabu) [reply specs/monitor_ceph_pool_utilization.adoc:172]: ok can you please add the component issues under References then

Comment (shtripat) [reply specs/monitor_ceph_pool_utilization.adoc:172]: Done

**Review by anmolbabu — APPROVED** [review]
LGTM but awaiting @brainfunked responses on "We are thinking of having some kind of tagging mechanism for the nodes and using that the leader mon can have this enabled.
@brainfunked what do you feel? Should we keep this requirement here at the moment or later we would take care"

**Comment by shtripat** [comment]
@anmolbabu https://github.com/Tendrl/specifications/issues/44 talks about this tag based job execution and its linking with role based job execution on node agents.

**Pull request merged**
